Added ann example, fresh rebase.

lesterlitch opened the pull request.

maciejkula commented:
Looks good. Can you please remove the `.DS_store` files?

lesterlitch commented:
That's done.

rth commented on examples/ann/annoy_nsmlib_example.ipynb:
```diff
@@ -0,0 +1,725 @@
+    "    \n",
+    "    #Pairwise\n",
+    "    \n",
+    "    data = np.repeat(item_vectors, i, axis=0)   \n",
```
Could try to use `numpy.broadcast_to` (see [this SO anwser](https://stackoverflow.com/a/37819585/1791279))

lesterlitch replied:
I can only see this used with 1d arrays? Looks like a useful function.

rth commented on examples/ann/annoy_nsmlib_example.ipynb:
```diff
@@ -0,0 +1,725 @@
+    "import numpy as np\n",
+    "\n",
+    "x = np.arange(10)\n",
+    "x = (x*10) * 1681\n",
```
It's not clear what's happening here. Shouldn't `x` be stored as the same time as the timing is computed?

rth commented on examples/ann/annoy_nsmlib_example.ipynb:
```diff
@@ -0,0 +1,725 @@
+    "\n",
+    "def time_nms_recommend_item_for_item():\n",
+    "    start = time.time()\n",
+    "    nms_idx.knnQuery(item_vectors[0])[0]\n",
```
You should explicitly pass `nms_idx`, `item_vectors` as inputs of this function, otherwise the `nms_index` that gets used is the one from several cells before not the one defined ~20 lines below.

Maybe also generally re-factor this whole section, to have setup functions for each algorithm,  make `response_times = {'annoy': [], 'nms': []..}` use a for loop over method name, and compute the timing in this second inner loop.

lesterlitch replied:
Yes much better.

benfred commented on examples/ann/annoy_nsmlib_example.ipynb:
```diff
@@ -0,0 +1,725 @@
+    "\n",
+    "Annoy has the advantage of saving to a memory map, and so playing well with multiple processes in a production environment\n",
+    "\n",
+    "NMSlib has generally been shown to outperform Annoy in terms of speed and accuracy, but I also found that constructing indexes over a large dataset was quite slow. This becomes a problem if you have to rebuild your index frequently because of fast moving product catalogues etc. "
```
Maybe try without the {'post':2} option for NMSLib - it roughly doubles the time to create a nmslib index. It does improve the recall rate, but its still possible to get a good QPS at over a 99% recall without it.   I have some breakdowns comparing the times with and without it here: https://github.com/benfred/bens-blog-code/tree/master/approximate_als.

Also I think you might want to note that Annoy is a pretty good choice for calculating similar items, but for some reason it isn't a great choice for recommending items after making that transform. I benchmarked this using Annoy/NMSLib/Faiss for recommending items here: http://www.benfrederickson.com/approximate-nearest-neighbours-for-recommender-systems/  - and its not clear that Annoy is better than just using a brute force solution for recommending items.

maciejkula commented:
These are all very good comments, but I feel quite bad about asking for so many changes!

@lesterlitch how would you prefer to handle this?

lesterlitch commented on examples/ann/annoy_nsmlib_example.ipynb:
```diff
@@ -0,0 +1,725 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We can get the item vectors by simply multiplying the item features by the models item feature embeddings:"
```
Done.

lesterlitch commented:
I've attempted to incorporate all the feedback so far. I've added a precision at k comparison with the ANN libraries, amended the variable names etc. and added a bit more explicit commentary. I've also tidied up the timing function to make it a single loop. Cheers.

maciejkula commented:
I like the changes, thanks for making them.

If I were in a nitpicky mood, I'd say that:
- the structure (headings/subheadings and so on) could still be a bit better,
- a lot of the lines are too long and wrap, we could make them shorter.

Any other comments @rth @benfred?

maciejkula commented on examples/ann/annoy_nsmlib_example.ipynb:
```diff
@@ -0,0 +1,780 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "<H1>Light FM Approximate Nearest Neighbours (ANN) example</H1>\n",
```
Did you actually write the HTML?

You should use [Markdown](http://jupyter-notebook.readthedocs.io/en/stable/examples/Notebook/Working%20With%20Markdown%20Cells.html) instead.

maciejkula commented on examples/ann/annoy_nsmlib_example.ipynb:
```diff
@@ -0,0 +1,780 @@
+    "\n",
+    "We'll look at two popular ANN libraries, Annoy and NMSlib, and compare their performance.\n",
+    "\n",
+    "<h4>Dependancies:</h4>\n",
```
Should be: dependencies

maciejkula commented:
I think this is a great example. I'm going to merge. Thanks all for your help!
